napari/config/core.py

```diff
@@ -0,0 +1,689 @@
+         ``config.paths`` list.  Notably, this includes
+         ``~/.config/napari/napari.yaml`` and anything else in there.
+      - ``collec_env()`` then looks for any environment variables that begin
+         with ``NAPARI_``
+   b. finally, ``collect()`` calls `merge()` on all of those discovered dicts,
+      to yield one merged result and returns it to the ``refresh()`` function
+      (by default, dicts discovered later in the process take priority).
+4. back in the ``refresh()`` function, that final merged result is used to
+   update the global config dict.
+5. At that point, the ``napari.yaml`` in the repo is read, if one doesn't
+   already exist, a file is created at ``~/.config/napari/napari.yaml`` (with
+   ``ensure_file()``) that is an exact copy of the repo ``napari.yaml`` but
+   with everything commented out, as a way to show users what they can change.
+6. finally, ``update_defaults()`` is called using the settings in the
+   `napari.yaml` file, which merges the configs again, but gives priority to
+   already existing values in the config (i.e. ones that were overwritten by
+   the user or the environement)
+
+Writing
+-------
+
+To write settings to disk, the :func:`sync` function is used.  This
+synchronizes the current ``config`` dict to a file at
+``join(config.PATH, 'napari.yaml')``.  If that file has changed since the last
+sync, it also reads new values from that file into the config.  If *both*
+the file and the config have changed, it prioritizes values in the config.
+"""
+import ast
+import logging
+import os
+import sys
+import threading
+from collections import defaultdict
+from functools import wraps
+from types import TracebackType
+from typing import (
+    Any,
+    Callable,
+    DefaultDict,
+    Dict,
+    List,
+    Optional,
+    Sequence,
+    Set,
+    Tuple,
+    Type,
+    Union,
+)
+
+import yaml
+
+from ..utils.misc import StringEnum
+from .deprecations import check_deprecations
+from .utils import canonical_name, ensure_file, merge, update
+
+#: a list of paths that get searched for .yaml files in :func:`collect_yaml`
+paths = [
+    os.getenv("NAPARI_ROOT_CONFIG", "/etc/napari"),
+    os.path.join(sys.prefix, "etc", "napari"),
+    os.path.join(os.path.expanduser("~"), ".config", "napari"),
+    os.path.join(os.path.expanduser("~"), ".napari"),
+]
+
+
+#: The primary directory for user config files
+#:
+#: Defaults to ~/.config/napari
+#: May be set with the environment variable "NAPARI_CONFIG"
+PATH = os.path.join(os.path.expanduser("~"), ".config", "napari")
+if "NAPARI_CONFIG" in os.environ:
+    PATH = os.environ["NAPARI_CONFIG"]
+    paths.append(PATH)
+
+#: a special yaml file to dump settings that are changed within a session
+#:
+#: this file will be loaded, but takes lower priority than all non-private
+#: yaml files.  So if a user has a specific setting in their napari.yaml file,
+#: it will override anything in _session.yaml
+_SESSION = os.path.join(PATH, '_session.yaml')
```

Comment:
The `_session.yaml` is still not quite clear to me, it seems like this is where things I changed during my session got saved, which is fine, but then would the "factory" defaults be in `napari.yaml`, or those are stored commented and eventually the user will make changes to `napari.yaml` too? Maybe when the explicitly hit some sort of "save preferences" or something? (i.e. this lets us capture both those save in the background things like window position AND the save when we ask sort of things too?) I certainly don't mind the file, just want to make sure I understand.

Reply:
- `napari.yaml`: factory defaults
- `~/.config/napari/napari.yaml`: user defaults ... get's copied (commented out) from `napari.yaml` on first run.  May be modified both programmatically and by the user.
- `_session.yaml` was where I was putting values that we do want to save from session to session, that are based on user interactivity, but which we don't necessarily want to be in `~/.config/napari/napari.yaml`.  Shouldn't be modified by the user.

An argument could be made that we just put _everything_ in the user config... but then the point of making it a human-readable yaml file starts to get a little lost.  There are almost certainly going to be values that we want to save, that we don't want a curious user stumbling upon and messing up state.  So think of `_session.yaml` as "private" settings, and `~/.config/napari/napari.yaml` as "public" settings.  I don't see it as depending that much on _when_ the preference is saved (i.e. hitting "save preferences" vs some background task) ... but rather _what_ is being saved (private setting vs public setting).

In a strict app ... all of this would be private, and the user would never look at the `config.yaml` file on disk: they would only interact with those settings in a GUI.
In a strict library, the user would be all over the `config.yaml` file ... using it to tune the library as desired... and the library probably wouldn't have any private settings like window position or "have we asked the user ____"

here, it's a bit confusing. #834  was a very private settings approach.  This one tends towards a very public settings approach.  Neither one feels 100% correct to me